### DISCUSSION TITLE
create a multi-environment package

### DISCUSSION THREAD

**Discussion by gidonk-pr** [serverless/serverless#10727] [Q&A]
Hi.

with older versions of the serverless framework we used to package the stack once (using `sls package`) and then use the same package to deploy the stack to different environments with slight variations, such as VPC configuration (different accounts different subnets etc.). We accomplished this by using env variables (or at some cases custom options) in the serverless.yaml file, changing the variable values per deployment (we remove the serverless-state.json file)

It seems that with newer versions, this is no longer possible, as now `sls package` requires the serverless-state.json file for the deployment of the package. This state file includes the environment variables **values** used at the time of packaging (rather than referencing the variables). serverless looks only in the state file, and disregards the serverless.yaml file, and is unable to detect any change in the variables value.

Recently (a few days ago) - there was a new update to the code, allowing to use `--param` flag to pass custom parameters to serverless. But the behaviour of this option seems to be the same as env variables when packaging.

This change seems peculiar to me for two main reasons:
1. serverless.yaml is still required even though the serverless-state.json file already contains all the required information that serverless.yaml file has. why the duplication?
2. while I understand that there are use cases for translating variable names into hard values in the serverless-state.json file, there are plenty of use cases for keeping variable references in the file as well. Having both environment variable and parameters behave the same, seems to be redundant to me. Instead having one (preferably environment variables) keep its reference in the state file, while the other (preferably the parameters) translated to hard values seems much more logical to me.

Is this behaviour by design or should I open a bug report? if it is by design- what would be the correct way to create a multi-environment serverless package? (with a nondeterministic number of environments)

**Comment by pgrzesik** [comment]
Hello @gidonk-pr, thanks for your report. What "older" versions are you referring to? For a long time the Framework `sls package` command works the same way as it's working right now. Regarding first question - The `serverless-state.json` file doesn't really introduce any extra information on top of what is already present in the resolved CloudFormation template files - I believe it was meant to serve as an extension for things that cannot be supported by CloudFormation, but it's not used for much at the moment. As for the second question - in order to actually perform deployment, we need resolved CloudFormation files and they depend on all resolved variables. The only thing that could potentially be reused across the deployments are the package artifacts (but these also can be influenced by configuration across stages). We already have an open issue for separating the packaging and deployment stages a bit, that should help out reusing e.g. already built zip artifacts for functions: https://github.com/serverless/serverless/issues/8499

**Reply by gidonk-pr** [reply]
Hi. thanks for the reply @pgrzesik 
We are using version 1.8.3 (I know- deprecated and old. we are working on upgrading, and is the reason for opening this thread)

We used to reused packages by referring to the zip file with the `--artifact` flag 
(I suppose we were following this guide for the most part: https://medium.com/@bishwash.aryal/serverless-framework-build-immutable-package-for-ci-cd-pipeline-949b49f7df91 )

Regarding the `serverless-state.json` file - I was playing around with package option quite a bit while trying to reuse it for different deployments. What I discovered is that the only files sls needs for the deployment are the generated zip files, the state file, and the `serverless.yaml` file. The cloudformation json files can be deleted (`cloudformation-template-create-stack.json`and `cloudformation-template-update-stack.json`) 

Moreover- the `serverless.yaml` file isn't really used. In my tests I completely changed the contents of the `serverless.yaml` file, and removed all contents of the file but the bare minimum serverless required for the deployment, which is the provider.name, and the service name, that I changed its value (the value that was used for the service name was what was saved in the state file)

**Reply by pgrzesik** [reply]
Hey @gidonk-pr - in v2 you can still use the same pattern as in the listed article, in v3 you will need to use a different way to supply the `--artifact` option as it will no longer be recognized. As for the other part - it's impossible to deploy without `cloudformation-template-create-stack.json` and `cloudformation-template-update-stack.json` - if you remove them and then run `deploy`, these will be generated again anyway because these are actually used for deployments. The state for `serverless-state.json` isn't really used for anything meaningful. If your deployment succeeded, please see what was actually deployed, because if you removed stuff from your configuration file, the results of your deployment will reflect what is defined in your configuration file.

**Reply by gidonk-pr** [reply]
Thanks for the suggestion of V2, we might revert to that version in the meanwhile until [#8499](https://github.com/serverless/serverless/issues/8499) is resolved. (is there an ETA or something that we can get an idea for when this will happen?)

Regarding the cloudformation-template files, I cannot positively determine if they are recreated on deploy- but they are definitely not recreated at the package path.
As for the `serverless.yaml` - I double checked myself, and must insist it's values are not used at all during deployment.

I am left a bit confused:
If the `serverless-state.json` file isn't used for anything meaningful, why is it required? 
Also- if the cloudformation files are recreated on deployment, where are the resolved values of the variables stored (that is- if they are not from the `serverless-state.json` file)?
From your description- it seems there isn't really anything preventing from sls from calculating the variables values again on deployment..

**Reply by pgrzesik** [reply]
Hey @gidonk-pr - there is not ETA for #8499. 

As for the CloudFormation templates - unless you're using some specific plugin that would change it (I'm not aware of such plugin), it's impossible to perform a deployment without using them so they're definitely recreated during `sls deploy` command. 

> Also- if the cloudformation files are recreated on deployment, where are the resolved values of the variables stored (that is- if they are not from the serverless-state.json file)?

They are not stored anywhere - they're resolved from your serverless configuration (`serverless.yml` or different file that you're using).

> If the serverless-state.json file isn't used for anything meaningful, why is it required?

I believe this is used for a minor piece of logic and that's why it's required, I believe there were bigger plans for this file but it's not really used for much at the moment. 

> From your description- it seems there isn't really anything preventing from sls from calculating the variables values again on deployment..

Nothing is preventing from doing that and it's actually what happens if you don't use `--package` flag. 

If you're observing different behavior, please include a small reproducible example so I can verify and reproduce on my side, otherwise it's hard to say more about your specific use case.

**Reply by gidonk-pr** [reply]
Hi.

I added an example of what I am doing: https://github.com/gidonk-pr/sls-package-test (using serverless version 3.2.1)

deploying the project as is (`sls deploy --region us-east-1 --stage dev -p ./package`) should work (at least for me it does) even though the `serverless.yaml` file is practically empty. From additional tests I made, changing the values in the state files will change the values deployed as well.

also- v2 has does not support the `--artifact` (and `--package` does not work the same as `--artifact`). seems to work the same as V3

**Reply by pgrzesik** [reply]
Hello @gidonk-pr - thanks a lot for that.

I'm honestly surprised and it seems like I was wrong - when `--package` is provided it actually relies on the `serverless-state.json` file to retrieve the CloudFormation template - sorry about the confusion here, I was not aware that it's actually relying on that. 

As for the `artifact` flag, in works in v2 in the same way as is described in the article, it won't work out of the box by just providing `--artifact` flag

**Reply by gidonk-pr** [reply]
We ended up re-packaging before every deployment, and then swapping out the zip files with the previously packaged file. it's a really convoluted way of doing it as we are forced to scan and replace checksum hashes in the state file.

**Reply by pgrzesik** [reply]
Thanks for sharing @gidonk-pr 🙇 So it seems like this https://github.com/serverless/serverless/issues/8499 issue should help out with the process, but as I mentioned there's currently no ETA for it :(
